### DISCUSSION TITLE
Disallowing extras

### DISCUSSION THREAD

**Discussion by kfsone** [tree-sitter/tree-sitter#1902] [Q&A] [answered]
Trying to write an expression to handle paths with wildcards in a grammar that uses '//' as comment separator and stumbling quite heavily on the meaning/use of token and token.immediate, which I expected to allow me to group sub-definitions without repeating them:

```
identifier: $ => /[A-Za-z_]\w+/,
variable_ref $ => token.immediate(seq('$', $.identifier)),
```
produces
```
Error processing rule variable_ref

Caused by:
    Grammar error: Unexpected rule Symbol(Symbol { kind: NonTerminal, index: 9 })
```

So I looked to the bash grammar as a guide, and whether it correctly enforced "x=1" for assignment, vs "x = 1", but it resorts to an external scanner.

Is there not an ability to disable extras-matching between non-terminal tokens? This undoes my delight at how tree-sitter implemented extras while still allowing you to explicitly match on those tokens.

What I was trying to do:

```
extras: $ => [
  $._comment,
  /\s/
],

_string_literal: $ => /"(\\[^\r\n]|""|[^\\"])*"/,

# paths may begin and/or end with a separator,
# but there must be at least one filename/wildcard component, thus
# invalid:
# /
# //...   <- comment
# .// 
# valid:
# .  *  .*.   file.name.***..(**).a
# /.  ./  /./  ./.  /./.  ././
_path_name:   $ => choice($._string_literal, /[A-Za-z0-9_\-\[\]\{\}\(\)\.\*]+/),
_path_step:     $ => seq($._path_name, '/'),
_path:             $ => seq(
  optional('/'), 
  choice(
    seq(repeat1($._path_step, optional($._path_name))),
    $._path_name
  )
)
```

Is there a way to do the above or is it better to not use extras and have explicit whitespace matches instead?

**Comment by ObserverOfTime** [comment, marked as answer]
Sometimes it is indeed better to not use extras and have explicit whitespace matches instead.
